packages/utilities/src/GlobalSettings.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * Storing global state in local module variables has issues when more than one copy
+ * if the module gets loaded on the page (due to a bundling error or simply by consuming
+ * a prebundled script.)
+ *
+ * This file contains helpers to deal with the getting and setting local state, and allows
+ * callers to get called back when it mutates.
+ */
+
+const GLOBAL_SETTINGS_PROP_NAME = '__globalSettings__';
+const CALLBACK_STATE_PROP_NAME = '__callbacks__';
+
+let _globalSettings: { [key: string]: any } = {};
+let _counter = 0;
+
+if (typeof window !== 'undefined') {
+  let win = window as any;
+
+  _globalSettings = win[GLOBAL_SETTINGS_PROP_NAME] = win[GLOBAL_SETTINGS_PROP_NAME] || {
+    [CALLBACK_STATE_PROP_NAME]: {}
+  };
+}
+
+const _callbacks = _globalSettings[CALLBACK_STATE_PROP_NAME];
+
+/**
+ * Change description used for change callbacks in GlobalSettings.
+ *
+ * @public
+ */
+export interface IChangeDescription {
+  key: string;
+  oldValue: any;
+  value: any;
+}
+
+/**
+ * Change event callback.
+ *
+ * @public
+ */
+export interface IChangeEventCallback {
+  __id__?: string;
+  (changeDescription?: IChangeDescription): void;
+}
+
+/**
+ * Global settings helper, which stores settings in the global (window) namespace.
+ * If window is not provided, it will store settings in module scope. Provides a
+ * way to observe changes as well when their values change.
+ *
+ * @public
+ */
+export class GlobalSettings {
+  public static getValue(key: string): any {
+    return _globalSettings[key];
+  }
+
+  public static setValue(key: string, value: any): void {
+    let oldValue = _globalSettings[key];
+
+    if (value !== oldValue) {
+      _globalSettings[key] = value;
+
+      let changeDescription = {
+        oldValue,
+        value,
+        key
+      };
+
+      for (let id in _callbacks) {
+        if (_callbacks.hasOwnProperty(id)) {
+          _callbacks[id](changeDescription);
+        }
+      }
+    }
+  }
+
+  public static addChangeListener(cb: IChangeEventCallback): void {
+    let id = cb.__id__;
+
+    if (!id) {
+      id = cb.__id__ = String(_counter++);
```

Comment:
It's a performance optimization to make a quick lookup table. This helps optimize add/remove listeners, as it's cheaper to delete from a map than it is to get index and splice an array.